### PR TITLE
Add resistance gene depth qc

### PR DESCRIPTION
**Pull request by taranewman**
solution for calculating gene coverage from #12 

Note: double check calculate_res_gene_depth.py is indexing the genome position correctly and confirm calculation is correct for one sample/gene before merging.

### CODE REVIEW

File: bin/calculate_res_gene_depth_qc.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    return (covered_positions / total_positions) * 100
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Calculate resistance gene coverage qc metrics.')
+    parser.add_argument('--bed', required=True, help='Path for resistance genes bed file.')
+    parser.add_argument('--depth', required=True, help='Path for tsv file with depth for each position (intermediate mpileup output).')
+    parser.add_argument('--threshold', type=float, default=10, help='min_depth threshold (default: 10).')
+    parser.add_argument('--output',  help='Output CSV file name.')
+    args = parser.parse_args()
```
Comment (dfornika) [inline bin/calculate_res_gene_depth_qc.py:37]: Can we move the argument parsing logic down below the `if __name__ == '__main__':` line, then pass the args into main? ie:

```python
def main(args):
    ...

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Calculate resistance gene coverage qc metrics.')
    parser.add_argument('--bed', required=True, help='Path for resistance genes bed file.')
    parser.add_argument('--depth', required=True, help='Path for tsv file with depth for each position (intermediate mpileup output).')
    parser.add_argument('--threshold', type=float, default=10, help='min_depth threshold (default: 10).')
    parser.add_argument('--output',  help='Output CSV file name.')
    args = parser.parse_args()
    main(args)
```

File: bin/calculate_res_gene_depth_qc.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    # Read gene positions from the resistance genes bed file
+    gene_data = []
+    with open(args.bed, 'r') as bed_file:
+        for line in bed_file:
+            fields = line.strip().split()
+            gene_name = fields[3]
+            start_position = int(fields[1])
```
Comment (dfornika) [inline bin/calculate_res_gene_depth_qc.py:45]: I think we may need to add 1 to the start position in the bed file to make it consistent with the start position in the depths file.

https://www.htslib.org/doc/samtools-mpileup.html#Pileup_Format
https://en.wikipedia.org/wiki/BED_(file_format)
https://en.wikipedia.org/wiki/General_feature_format

File: bin/calculate_res_gene_depth_qc.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            depth_at_position[position] = depth
+
+    # Calculate qc metrics and write results to csv
+    output_file = args.output
+    with open(output_file, 'w', newline='') as csv_file:
+        writer = csv.writer(csv_file)
+        writer.writerow(['gene_name', 'gene_position_start', 'gene_position_end', 'mean_depth_coverage', 'percent_of_gene_covered_above_depth_threshold'])
+        
+        for gene_name, gene_positions in gene_data:
+            mean_depth = calculate_mean_depth(gene_positions, depth_at_position)
```
Comment (dfornika) [inline bin/calculate_res_gene_depth_qc.py:75]: Can we round the mean depth to 3 decimal places before outputting?

**Review by dfornika — CHANGES_REQUESTED** [review]
Looking great. Just a few small changes.

File: bin/calculate_res_gene_depth_qc.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import csv
```
Comment (dfornika) [inline bin/calculate_res_gene_depth_qc.py:1]: Can you add:

```
#!/usr/bin/env python3
```

...to the top of this script?

**Review by dfornika — CHANGES_REQUESTED** [review]
Sorry, one more change. Can you add a 'shebang' line to the top of the script?

```
#!/usr/bin/env python3
```

File: bin/calculate_res_gene_depth_qc.py (hunk 1 above)
Comment (taranewman) [reply bin/calculate_res_gene_depth_qc.py]: [Fixed here](https://github.com/BCCDC-PHL/tbprofiler-nf/pull/14/commits/067b0f6dc533901d7dc196972d189136e8d62de1)

File: bin/calculate_res_gene_depth_qc.py (hunk 4 above)
Comment (taranewman) [reply bin/calculate_res_gene_depth_qc.py:3]: Good catch! [Thank you! ](https://github.com/BCCDC-PHL/tbprofiler-nf/pull/14/commits/067b0f6dc533901d7dc196972d189136e8d62de1)

File: bin/calculate_res_gene_depth_qc.py (hunk 2 above)
Comment (taranewman) [reply bin/calculate_res_gene_depth_qc.py]: Sounds good, fixed [here](https://github.com/BCCDC-PHL/tbprofiler-nf/pull/14/commits/59677233eed66f1bd191730946e9f94c381a1b97)

**Review by dfornika — APPROVED** [review]

**Pull request merged**